### ISSUE TITLE
Disabled webhooks periodically getting created after update to v3.3.4

### ISSUE THREAD

**Issue by achadee** [woocommerce/woocommerce#19634]
Hi guys, hope you are well

We have a customers woocommerce constantly creating webhooks after the version update. At first I thought it was a plugin, but we deactivated all the plugins and it is still creating webhooks.

I also tried setting our integration api keys to READ only. (that should stop it if it was the integration causing the issue?)

<img width="571" alt="screen shot 2018-04-06 at 2 39 03 pm" src="https://user-images.githubusercontent.com/5952918/38403761-17202eb4-39a9-11e8-8b20-33215a6bd11b.png">

I don't have access to the db, so can't see which user is creating the webhooks. Is there a way to tell where the webhooks are being created from?

## Prerequisites

- [x ] I have searched for similar issues in both open and closed tickets and cannot find a duplicate
- [ ] The issue still exists against the latest `master` branch of WooCommerce on Github (this is **not** the same version as on WordPress.org!)
- [ ] I have attempted to find the simplest possible steps to reproduce the issue
- [ ] I have included a failing test as a pull request (Optional)

## Steps to reproduce the issue
Cannot reproduce, Looking for advice in what might be causing the issue

- [x] This bug happens with only WooCommerce plugin active
- [ ] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/)
- [ ] I can reproduce this bug consistently using the steps above

## WordPress Environment

<details>
```
`
### WordPress Environment ###

Home URL: https://www.stepaheadshoes.com.au
Site URL: https://www.stepaheadshoes.com.au
WC Version: 3.3.4
Log Directory Writable: ✔
WP Version: 4.9.4
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: ✔
WP Cron: ✔
Language: en_US

### Server Environment ###

Server Info: LiteSpeed
PHP Version: 7.0.29
PHP Post Max Size: 256 MB
PHP Time Limit: 300
PHP Max Input Vars: 3000
cURL Version: 7.54.0
OpenSSL/1.0.2k

SUHOSIN Installed: –
Max Upload Size: 128 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 3.3.4
WC Database Prefix: wp_
MaxMind GeoIP Database: ✔
Total Database Size: 86.78MB
Database Data Size: 73.90MB
Database Index Size: 12.88MB
wp_woocommerce_sessions: Data: 0.08MB + Index: 0.00MB
wp_woocommerce_api_keys: Data: 0.00MB + Index: 0.01MB
wp_woocommerce_attribute_taxonomies: Data: 0.00MB + Index: 0.01MB
wp_woocommerce_downloadable_product_permissions: Data: 0.00MB + Index: 0.00MB
wp_woocommerce_order_items: Data: 0.08MB + Index: 0.05MB
wp_woocommerce_order_itemmeta: Data: 0.57MB + Index: 0.41MB
wp_woocommerce_tax_rates: Data: 0.00MB + Index: 0.00MB
wp_woocommerce_tax_rate_locations: Data: 0.00MB + Index: 0.00MB
wp_woocommerce_shipping_zones: Data: 0.00MB + Index: 0.00MB
wp_woocommerce_shipping_zone_locations: Data: 0.00MB + Index: 0.01MB
wp_woocommerce_shipping_zone_methods: Data: 0.00MB + Index: 0.00MB
wp_woocommerce_payment_tokens: Data: 0.00MB + Index: 0.00MB
wp_woocommerce_payment_tokenmeta: Data: 0.00MB + Index: 0.01MB
wp_woocommerce_log: Data: 0.00MB + Index: 0.00MB
wp_commentmeta: Data: 0.05MB + Index: 0.02MB
wp_comments: Data: 0.50MB + Index: 0.21MB
wp_cptch_images: Data: 0.02MB + Index: 0.00MB
wp_cptch_packages: Data: 0.02MB + Index: 0.00MB
wp_cptch_whitelist: Data: 0.02MB + Index: 0.02MB
wp_failed_jobs: Data: 0.00MB + Index: 0.00MB
wp_hctpc_images: Data: 0.02MB + Index: 0.00MB
wp_hctpc_packages: Data: 0.02MB + Index: 0.00MB
wp_hctpc_whitelist: Data: 0.02MB + Index: 0.02MB
wp_links: Data: 0.00MB + Index: 0.00MB
wp_linksync_laidKey: Data: 0.00MB + Index: 0.00MB
wp_linksync_log: Data: 33.47MB + Index: 0.48MB
wp_mailchimp_carts: Data: 0.03MB + Index: 0.00MB
wp_options: Data: 3.55MB + Index: 0.31MB
wp_postmeta: Data: 21.19MB + Index: 8.46MB
wp_posts: Data: 3.05MB + Index: 0.84MB
wp_queue: Data: 0.01MB + Index: 0.00MB
wp_revslider_css: Data: 0.09MB + Index: 0.00MB
wp_revslider_layer_animations: Data: 0.00MB + Index: 0.00MB
wp_revslider_navigations: Data: 0.00MB + Index: 0.00MB
wp_revslider_sliders: Data: 0.04MB + Index: 0.00MB
wp_revslider_slides: Data: 0.29MB + Index: 0.00MB
wp_revslider_static_slides: Data: 0.00MB + Index: 0.00MB
wp_smush_dir_images: Data: 0.00MB + Index: 0.00MB
wp_termmeta: Data: 0.00MB + Index: 0.01MB
wp_terms: Data: 0.00MB + Index: 0.01MB
wp_term_relationships: Data: 0.22MB + Index: 0.41MB
wp_term_taxonomy: Data: 0.00MB + Index: 0.01MB
wp_usermeta: Data: 1.59MB + Index: 1.10MB
wp_users: Data: 0.15MB + Index: 0.12MB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB
wp_wc_webhooks: Data: 1.52MB + Index: 0.25MB
wp_woof_query_cache: Data: 0.02MB + Index: 0.02MB
wp_wpml_mails: Data: 7.25MB + Index: 0.02MB
wp_yith_wcwl: Data: 0.01MB + Index: 0.01MB
wp_yith_wcwl_lists: Data: 0.00MB + Index: 0.02MB
wp_yoast_seo_links: Data: 0.00MB + Index: 0.00MB
wp_yoast_seo_meta: Data: 0.00MB + Index: 0.01MB

### Post Type Counts ###

attachment: 5829
cms_block: 4
mc4wp-form: 2
nav_menu_item: 47
page: 19
portfolio: 3
post: 2
product: 487
product_variation: 4411
revision: 241
shop_order: 690
shop_order_refund: 16
wpcf7_contact_form: 1

### Security ###

Secure connection (HTTPS): ✔
Hide errors from visitors: ✔

### Active Plugins (3) ###

WooCommerce Product Filter: by Mihajlovic Nenad – 6.5.1 – Not tested with the active version of WooCommerce
WooCommerce Stripe Gateway: by WooCommerce – 4.0.7
WooCommerce: by Automattic – 3.3.4

### Settings ###

API Enabled: ✔
Force SSL: –
Currency: AUD ($)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)


### WC Pages ###

Shop base: #458 - /shop/
Cart: #459 - /cart/
Checkout: #460 - /checkout/
My account: #461 - /my-account/
Terms and conditions: #468 - /shopping-made-easy/returns-and-exchanges/

### Theme ###

Name: Basel Child
Version: 1.0.0
Author URL: http://xtemos.com
Child Theme: ✔
Parent Theme Name: Basel
Parent Theme Version: 4.0.0
Parent Theme Author URL: http://themeforest.net/user/xtemos
WooCommerce Support: ✔

### Templates ###

Overrides: basel/woocommerce/archive-product.php
basel/woocommerce/cart/cart.php
basel/woocommerce/cart/cross-sells.php
basel/woocommerce/cart/mini-cart.php
basel/woocommerce/checkout/form-checkout.php
basel/woocommerce/content-product.php
basel/woocommerce/content-product_cat.php
basel/woocommerce/content-single-product.php
basel/woocommerce/global/breadcrumb.php
basel/woocommerce/global/form-login.php
basel/woocommerce/global/quantity-input.php
basel/woocommerce/loop/loop-end.php
basel/woocommerce/loop/loop-start.php
basel/woocommerce/loop/no-products-found.php
basel/woocommerce/loop/orderby.php
basel/woocommerce/loop/pagination.php
basel/woocommerce/loop/sale-flash.php
basel/woocommerce/myaccount/form-login.php
basel/woocommerce/myaccount/navigation.php
basel/woocommerce/product-searchform.php
basel/woocommerce/single-product/add-to-cart/variable.php
basel/woocommerce/single-product/meta.php
basel/woocommerce/single-product/product-image.php
basel/woocommerce/single-product/product-thumbnails.php
basel/woocommerce/single-product/related.php
basel/woocommerce/single-product/sale-flash.php
basel/woocommerce/single-product/tabs/tabs.php
basel/woocommerce/single-product/title.php
basel/woocommerce/single-product/up-sells.php
basel/woocommerce/single-product.php

`
```
</details>

**Comment by mikejolley** [comment]
That would disable them via our API but there are other ways to authenticate. The update script should only run once also which converts webhooks to the new post type format. Could that be running more times? Maybe the update routine is stuck?

See if the update log in system status > tools is running multiple times. Then you'll probably want to delete those empty webhooks so they don't stick around.

**Comment by mikejolley** [comment]
What makes you say it's creating new webhooks? Those log to a file.

**Comment by mikejolley** [comment]
To narrow it down you could:

- exit here - do they stop being created? https://github.com/woocommerce/woocommerce/blob/2aa74a6f66a410c4438c988d2cb004cca3bd3e99/includes/api/v1/class-wc-rest-webhooks-controller.php#L318 If so you know something is hitting your API.
- exit here - do they stop being created? https://github.com/woocommerce/woocommerce/blob/75fce4a5fa620bc9fa810a079b25cb79de0edef8/includes/wc-update-functions.php#L1504 If so you know the update routine is at fault and creating new webhook entries.

The webhooks are inside `wc_webhooks` table as of 3.3- maybe something is coded to create entries there also? The only code in WooCommerce which interacts with this table is `WC_Webhook_Data_Store`.

Let us know what you find. Unfortunately without steps to reproduce this I don't see how we can solve, nor can I see any code in core that would create webhooks in this manner. I can only assume it's rogue custom code somewhere; those webhooks don't even have names/statuses/valid data. To this end I'm closing the issue for now.

**Comment by achadee** [comment]
The customers dev team looked at it. Im not sure if this is helpful to anyone, but apparently there was a "pending" webhook stuck when they put their store in maintenance mode. And when the store was re-enabled the webhook was triggering constantly. Im still not entirely sure why it was repetitively creating webhooks, but the issue is fixed :)

**Comment by mikejolley** [comment]
cc @claudiosanches

**Comment by claudiosanches** [comment]
Maintenance mode probably will avoid to trigger webhooks, but still not sure what should be creating new webhooks, at least as far I know, this is the first time we saw something like that.

**Comment by achadee** [comment]
We have an issue where the defective webhooks are still triggering after I deleted all of them in the db.

Do you guys store a queue somewhere of all pending webhooks? Maybe in the database? or in memory?

**Comment by bobbiejwilson** [comment]
The same issue is happening on a site I manage. How do you find a pending webhook and shut it down? I've disabled the API, removed all apps and webhooks.

**Comment by bobbiejwilson** [comment]
And after expanding the details, I realize that this is the site I'm working on ha!

**Comment by bobbiejwilson** [comment]
@mikejolley After moving to a more supportive host, we were able to look into the cron jobs that were running. Whoever uninstalled a plugin the site was using to send out info, left that job running with a deleted webhook as an arg inside it. The cron event in particular is woocommerce_deliver_webhook_async. Can you confirm that this cron should be running (when I removed it the webhooks stopped firing so pretty sure)?

**Comment by claudiosanches** [comment]
By default `woocommerce_deliver_webhook_async` is a one time event, triggered with `wp_schedule_single_event`, so after deleted should not be triggered more times.

**Comment by bobbiejwilson** [comment]
Excellent. I don't see it in the db now, but it was definitely what was stuck this whole time. Hopefully this helps someone else.

**Comment by arwinvdv** [comment]
After deleting the 'cron' option in the wp_options table the same problem was fixed for me.

**Comment by XedinUnknown** [comment]
A client of mine had a similar problem. The solutions suggested here and in other places appeared to be inapplicable or ineffective. The common denominator, however, was that WooCommerce was unable to run for at least some time. While still not sure what the cause of the problem is, you can break the cycle of new jobs being scheduled like this:

- Shut down the web server and all PHP processes started by it.
- Re-start the database server if possible.
- Clear the jobs from the `cron` option.
- Start the web server back up.

**Comment by thenbrent** [comment]
For others coming to this issue, this PR will mitigate the damage of this issue: https://github.com/woocommerce/woocommerce/pull/21427

It doesn't solve the root issue of rogue webhooks being created (that will need to be resolved in whatever plugin or custom code is creating those webhooks), but it will prevent _all_ of those disabled, corrupted webhooks being loaded.

**Comment by lkraav** [comment]
(3.5.0, for safety)

Having now run into this very issue myself, it seems to be pretty damn difficult to debug where exactly these webhooks are being created from.

I noticed a few thousand pending action scheduler events, a lot of them on hook `woocommerce_deliver_webhook_async` and a webhook ID that didn't exist anymore.

To keep platform alive, I went in heavy-handed via `wp post delete $(wp post list --post_type=scheduled-action --post_status=pending --title=woocommerce_deliver_webhook_async --format=ids) --force` for time being.

We did deploy a single custom webhook topic callback solution a little while ago. Not sure if it's the cause yet (disabled for monitoring period), but if is, I never imagined it could fail dirty like this.

**Comment by Jeppeskovsgaard** [comment]
I had exactly the same issue with 10.000+ webhooks and about 150.000+ scheduled actions labeled as `woocommerce_deliver_webhook_async`. I don't know why that cron job wen't rogue, but after deleting all scheduled actions and webhooks in the database, it has stopped. At least in my case. 

Remember to also delete scheduled action_log entries in wp_comments. WooCommerce is querying wp_comments with an expensive query on every single control panel page (this is still an issue https://core.trac.wordpress.org/ticket/31072), so a large wp_comments table can really slow down your site. The query took about 1,3s to run on my site with 50.000+ action_log entries in the wp_comments table.

**Comment by robwatson-intechra** [comment]
I wonder if anyone here was also running MailChimp for Wordpress like I was. That's what caused the "thousands of webhooks" issue for me. I had installed MailChimp for WordPress over a month ago and had a huge spike in activity. MailChimp told me that was normal as it was doing some processing over my customer list to do its analytics. But after the processing was done, my server was taking 3 seconds per PHP transaction. We couldn't figure out why this was happening. It was kind of by accident that I discovered the Webhooks connection. Last night I was working another issue and happened to click on WooCommerce > Settings > Advanced > Webhooks and noticed well over 4,000 disabled Webhooks sitting there. I spent the next hour deleting them by hand, 20 at a time. Immediately, I regained the performance I had prior to installing MailChimp for WooCommerce. I'm 99.9% certain that MailChimp for WooCommerce created the rogue Webhooks.
